Homework 23

matt-pierce-test opened the pull request.

mattpierce017 committed:
Updated BaseTest

mattpierce017 committed:
Updating BasePage and setting up files and pages

mattpierce017 committed:
initialized BasePage

mattpierce017 committed:
updated LoginPage

mattpierce017 committed:
setting up Reusable elements in BasePage

mattpierce017 committed:
initialized pages in HW-23

mattpierce017 committed:
renamed AllSongsPage and uploaded tests

mattpierce017 committed:
Added methods to AllSongsPage

mattpierce017 committed:
refactored tests to POM for SongsTests

mattpierce017 committed:
Update HomePage.java

updated getAvatarIcon();

mattpierce017 committed:
working on LoginTests

mattpierce017 committed:
updated HomePage with editPlaylistName()

mattpierce017 committed:
refactored code for addSongToPlaylist

mattpierce017 committed:
Update SongsTests.java

mattpierce017 committed:
Update HomePage.java

mattpierce017 committed:
Updated TestNG.xml

mattpierce017 committed:
Working on running tests

mattpierce017 committed:
Page Factory

mattpierce017 committed:
refactored code

mattpierce017 committed:
updated locators. ran clear cache in IntelliJ. Still can't run tests.

Berichicko commented:
The code is not compiled! Please finish the homework!You got this problem, please correct it. 

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/b22d6011-3f2d-47f1-a287-82ae6db827d9)

HINT

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/e4869be2-486a-4527-99aa-3239863bba7d)


```
    public WebElement getAvatarIcon() {

       return avatarIcon;
    }
```

Berichicko commented on src/test/java/BaseTest.java:
```diff
@@ -1,11 +1,166 @@
+        };
+    }
+
+    public static void openLoginUrl() {
+        String url = "https://qa.koel.app/";
+        driver.get(url);
+    }
+
+    protected static void enterEmail(String email) throws InterruptedException{
+        WebElement emailInput = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("[type='email']")));
+        emailInput.click();
+        emailInput.clear();
+        //inputs valid email address into email input field
+        emailInput.sendKeys(email);
+    }
+
+    protected static void enterPassword(String password) throws InterruptedException{
+
+        WebElement passwordInput = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("[type='password']")));
+        passwordInput.clear();
+        //inputs valid password for login into password field
+        passwordInput.sendKeys(password);
+    }
+
+    public void clickSubmit() throws InterruptedException{
+        WebElement loginButton = wait.until(ExpectedConditions.elementToBeClickable((By.cssSelector("[type='submit']"))));
+        loginButton.click();
+    }
```
Page Object Model (POM): recommended to associate the methods with the corresponding page class where they are relevant. In the POM, each page or section of a web application is represented by a separate class. These classes encapsulate the elements and actions specific to that page or section.

Move these methods to the corresponding class according to the POM.

Berichicko commented on src/test/java/pages/HomePage.java:
```diff
@@ -0,0 +1,109 @@
+    }
+    public void getDeletionText(){
+
+        confirmationMessage.getText();
```
You got this problem, please correct it. 

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/2ed9f17c-83e4-40af-a6be-4648e7eae7d7)

HINT

Add 

```
   @FindBy(xpath = "[//div[@class='alertify-logs top right']")
    private WebElement confirmationMessage;
```
![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/42f152ae-8bbb-482a-9729-b1455fde97be)

Berichicko commented:
1. **BasePage and BaseTest** 

**BaseTest** — Implements the necessary methods for work with webdriver.

The BaseTest class contains common setup and teardown methods, as well as any shared functionality or configurations required by multiple test classes.

**BaseTest**

The problem we are trying to solve here is the reuse of common behaviors across test classes avoiding coding duplication and centralizing these actions at one point.

**@BeforeSuite**	- The annotated method will be run before all tests in this suite have run.

**@BeforeMethod** -	The annotated method will be run before all the test methods in the current class have been run.

**@AfterMethod**	- The annotated method will be run after each test method.

_Example without a Base Test Class_

![image](https://user-images.githubusercontent.com/104761482/233972627-2d7ffdc7-6a2f-4e9f-8a3a-a1b4a7a47bc4.png)


**BasePage**  — class serves as a foundation or parent class for individual page classes. It contains common methods and functionalities that are shared across multiple pages in an application.

_Example without a Base Page Class_

![image](https://user-images.githubusercontent.com/104761482/236891493-7f7013c7-fde2-4bf9-9084-3aa7483df8e3.png)


2. **Test in Page Object Model (POM)**

![Clean_Test](https://user-images.githubusercontent.com/104761482/236863813-f30d7bfc-e89d-4419-bdea-4b1517155e39.png)

3.  **Page Object Model and Page Factory**

![POM_vs_Factory_1](https://user-images.githubusercontent.com/104761482/236864030-34b89652-460d-4440-aa91-894332e90654.png)

![POM_vs_Factory_2](https://user-images.githubusercontent.com/104761482/236864098-d3ffe636-7ba9-484e-a72f-8b03ff024659.png)

4.  Please, use **Encapsulation**

Encapsulation is one of the four fundamental OOP concepts. The other three are inheritance, polymorphism, and abstraction.

You should use encapsulation of non-static and non-final fields.

If a variable is used only by other methods of that class, it should be declared as private.

Make all the locators and WebElements private, because private keyword in java allows most restrictive access to variables and methods and offer strongest form of Encapsulation. 

For example

    ```
    private By shuffleBtnLocator = By.cssSelector(".btn-shuffle-all");
    private By firstSongLocator = By.cssSelector(".play");

```
   @FindBy(css = "img.avatar")
    private WebElement avatar
```


Make the variables protected, because protected access gives the subclass a chance to use the helper method or variable, while preventing a nonrelated class from trying to use it.

```
    protected Actions actions;
    protected WebDriverWait wait;
    protected WebDriver driver;
```

6. Try to analyse status tests

_For example_

![Test status_en](https://user-images.githubusercontent.com/104761482/236867158-08437918-56e2-462b-b7c9-5629d9f16e74.png)

![Test failed_en](https://user-images.githubusercontent.com/104761482/236867187-6613ad47-d9a9-4d85-a1cf-3ca1ecf8a347.png)


![Test error_en](https://user-images.githubusercontent.com/104761482/236867225-898c86e0-89b6-4eb1-a707-1de6cae472d4.png)

mattpierce017 committed:
Updated format for POM

Should I get rid of the BaseTest completely since I have moved the methods?

I also am having a hard time with the IDE finding elements that I used in previous HWs

Berichicko commented on src/test/java/LoginTests.java:
```diff
@@ -1,25 +1,19 @@
-        WebDriver driver = new ChromeDriver(options);
-        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
+        loginPage.login();
+        homePage.waitForOverlay(overlayLocator);
```
The code is not compiled! Please finish the homework!You got this problem, please correct it. 

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/f2ee2a95-bcc2-440f-b817-6890c06c39e0)

HINT

**Step 1**

**public class BasePage**

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/ee7642ae-4e29-45bb-86ac-b703285d9c21)

    public  void waitForOverlayElement (WebElement element) {
        wait.until(ExpectedConditions.visibilityOf(element));
    }
This is a Java method that uses Selenium Webdriver's ExpectedConditions class to wait until a certain WebElement becomes visible on the webpage. The wait object used is an instance of WebDriverWait, which is initialized with a timeout value. The visibilityOf method of the ExpectedConditions class waits until the specified element becomes visible on the page, at which point the wait.until() method returns control to the caller. This method is commonly used to ensure that a webpage has finished loading an element or overlay before continuing with further actions.

**Step 2**

**public class HomePage**

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/a49f7c79-bd29-44f6-a48c-6a5dbce2aa29)

 Java method that returns a WebElement representing an avatar or profile picture icon on a webpage. The method uses the waitForOverlayElement method, which we've previously discussed, to ensure that the avatar icon is visible on the page before returning it.
 
 **Step 3**
 
 ![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/94ec9635-bfb4-4ce5-995b-8b9ea199b59b)
 
 
![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/662eb5e9-193b-43f9-854e-8acbd79a8b62)

mattpierce017 committed:
refactored login(), playSong(), and createPlaylist()

mattpierce017 committed:
fixed issues

Berichicko commented on src/test/java/ProfileTests.java:
```diff
@@ -0,0 +1,43 @@
+public class ProfileTests extends BaseTest {
+
+    @Test
+    public void updateProfileName () throws InterruptedException {
```
Please, delete `throws InterruptedException`

Thread. sleep is bad! It blocks the current thread and renders it unusable for further work.
Using Thread. sleep() frequently in an automation framework is not a good practice.

Please use the Explicit Wait command, the WebDriver is directed to wait until a certain condition occurs before proceeding with executing the code.

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/b489b4a3-afeb-4b60-a3a7-be14e0bec20e)

mattpierce017 committed:
removed sleep

Berichicko commented:
Hello,

I have checked your homework 23, and I wanted to let you know that you're doing great! Keep up the excellent work!
Best wishes from the HW team (Alex, [aliaksandr@testpro.io](mailto:aliaksandr@testpro.io))!

![804deb64dc406a20a75da05d91a1a69e](https://user-images.githubusercontent.com/104761482/235961479-9afb2841-b657-46c9-99f8-2402f933b843.jpg)

![image](https://github.com/Test-Pro-Bootcamp/A47-QA-Automation-Framework/assets/104761482/0415da4d-07eb-42e7-8345-e7513fed18f5)
